gusto/linear_solvers.py

```diff
@@ -220,6 +220,182 @@ def solve(self):
+                           'fieldsplit_1_pc_type': 'gamg',
+                           'fieldsplit_1_mg_levels_ksp_type': 'chebyshev',
+                           'fieldsplit_1_mg_levels_ksp_chebyshev_estimate_eigenvalues': True,
+                           'fieldsplit_1_mg_levels_ksp_chebyshev_estimate_eigenvalues_random': True,
+                           'fieldsplit_1_mg_levels_ksp_max_it': 1,
+                           'fieldsplit_1_mg_levels_pc_type': 'bjacobi',
+                           'fieldsplit_1_mg_levels_sub_pc_type': 'ilu'}
+        else:
+            self.params = params
+
+        # setup the solver
+        self._setup_solver()
+
+    def _setup_solver(self):
+        state = self.state      # just cutting down line length a bit
+        dt = state.timestepping.dt
+        beta = dt*state.timestepping.alpha
+        cp = state.parameters.cp
+        mu = state.mu
+        Vu = state.spaces("HDiv")
+        Vu_broken = FunctionSpace(BrokenElement(Vu.ufl_element()))
+        Vtheta = state.spaces("HDiv_v")
+        Vrho = state.spaces("DG")
+        HOW DO WE GET THE TRACE SPACE IN THE EXTRUDED CASE?
+        Vtrace = FunctionSpace(mesh, "HDiv Trace", tdegree)
+
+        # Split up the rhs vector (symbolically)
+        u_in, rho_in, theta_in = split(state.xrhs)
+
+        # Build the reduced function space for u,rho
+        M = MixedFunctionSpace((Vu_broken, Vrho, Vtrace))
+        w, phi, dl = TestFunctions(M)
+        u, rho, l0 = TrialFunctions(M)
+
+        n = FacetNormal(state.mesh)
+
+        # Get background fields
+        thetabar = state.fields("thetabar")
+        rhobar = state.fields("rhobar")
+        pibar = exner(thetabar, rhobar, state)
+        pibar_rho = exner_rho(thetabar, rhobar, state)
+        pibar_theta = exner_theta(thetabar, rhobar, state)
+
+        # Analytical (approximate) elimination of theta
+        k = state.k             # Upward pointing unit vector
+        theta = -dot(k,u)*dot(k,grad(thetabar))*beta + theta_in
+
+        # Only include theta' (rather than pi') in the vertical
+        # component of the gradient
+
+        # the pi prime term (here, bars are for mean and no bars are
+        # for linear perturbations)
+
+        pi = pibar_theta*theta + pibar_rho*rho
+
+        # vertical projection
+        def V(u):
+            return k*inner(u,k)
+
+        # specify degree for some terms as estimated degree is too large
+        dxp = dx(degree=(self.quadrature_degree))
+        dS_vp = dS_v(degree=(self.quadrature_degree))
+        dS_hp = dS_v(degree=(self.quadrature_degree))
+
+        eqn = (
+            inner(w, (u - u_in))*dx
+            - beta*cp*div(theta*V(w))*pibar*dxp
+            # following does nothing but is preserved in the comments
+            # to remind us why (because V(w) is purely vertical.
+            # + beta*cp*jump(theta*V(w),n)*avg(pibar)*dS_v
+            - beta*cp*div(thetabar*w)*pi*dxp
+            + beta*cp*jump(thetabar*w,n)*l0*(dS_vp + dS_hp)
```

Comment:
From my previous comment above about the mixed space, now for `eqn` you only need to write the forms corresponding to `w, u` and `phi, rho`. So just drop the `l0, dl` terms as they will be handled later.